clang/lib/Serialization/ASTReader.cpp

```diff
@@ -3349,10 +3351,29 @@ ASTReader::ReadControlBlock(ModuleFile &F,
                                       .getModuleCache()
                                       .getInMemoryModuleCache()
                                       .isPCMFinal(F.FileName);
-      if (isDiagnosedResult(Result, Capabilities) || recompilingFinalized)
+      if (isDiagnosedResult(Result, Capabilities) || recompilingFinalized) {
         Diag(diag::note_module_file_imported_by)
             << F.FileName << !F.ModuleName.empty() << F.ModuleName;
 
+        // When a Filemap is used, the ImportedFile may be different than what
+        // was encoded as a dependency in StoredFile.
+        // When this happens, look up both files when reporting that they
+        // differ. If either file does not exist, fall back to a string path
+        // check.
+        if (UseFilemap && !F.ModuleName.empty() && !StoredFile.empty()) {
+          auto ImportedFileRef =
+              PP.getFileManager().getOptionalFileRef(ImportedFile);
+          auto StoredFileRef =
+              PP.getFileManager().getOptionalFileRef(StoredFile);
+          const bool MissingFiles = !ImportedFileRef || !StoredFileRef;
+          if ((MissingFiles && (StoredFile != ImportedFile)) ||
```

Comment:
If either file is missing, haven't we already emitted a diagnostic elsewhere? I'm thinking this might only add noise to the previous diagnostic (complaining about benign path differences).

Reply:
I don't think so. `StoredFile` is never checked for existence anywhere else &  this code path can only happen if `ImportedFile` existed at the time of `lookupModuleFile`. 

It would be nice to reuse the FileEntry created by the ModuleManager, but it goes out of scope by the time this check happens. The comment in https://github.com/llvm/llvm-project/blob/main/clang/lib/Serialization/ModuleManager.cpp#L449 also implies to me that it's possible the file could go _missing_ between those calls too. 

Somewhat related tangent: I'd like to rework a lot of this code so that when an `OutOfDate` case happens, the `ImportedFile` pcm is opened and we can report more user-facing information that represents the size check (e.g. modulemaps conflict), but that's out of scope right now.

Reply:
> `StoredFile` is never checked for existence anywhere else

That makes sense.

> & this code path can only happen if `ImportedFile` existed at the time of `lookupModuleFile`.

Then `FileManager` will never report it as missing after that point, unless I'm missing something.

> It would be nice to reuse the FileEntry created by the ModuleManager, but it goes out of scope by the time this check happens.

Does it? It seems like it gets stored as `ModuleFile::File`.

Reply:
> > It would be nice to reuse the FileEntry created by the ModuleManager, but it goes out of scope by the time this check happens.
> 
> Does it? It seems like it gets stored as `ModuleFile::File`.

Ah, we're talking about the case where `lookupModuleFile()` fails. But I still don't see how the file can go missing if it's already stored in the `FileManager` cache. That comment specifically talks about first calling with `/*OpenFile=*/false` and then `/*OpenFile=*/true`, which I don't think you care about.